opal/mca/btl/sm/btl_sm_module.c

```diff
@@ -194,18 +194,30 @@ static int sm_btl_first_time_init(mca_btl_sm_t *sm_btl, int n)
 }
 
 
-static int init_sm_endpoint (struct mca_btl_base_endpoint_t *ep, struct opal_proc_t *proc, int remote_rank) {
+static int init_sm_endpoint (struct mca_btl_base_endpoint_t **ep_out, struct opal_proc_t *proc) {
     mca_btl_sm_component_t *component = &mca_btl_sm_component;
     union sm_modex_t *modex;
     ino_t my_user_ns_id;
     size_t msg_size;
     int rc;
 
+    uint16_t peer_local_rank;
+    uint16_t *ptr = &peer_local_rank;
+    OPAL_MODEX_RECV_VALUE(rc, PMIX_LOCAL_RANK, &proc->proc_name,
```

Comment:
@rhc54 PMIX_LOCAL_RANK has been around since the beginning, right?  Trying to remember the details of our conversations last year around all the different rankings that we available when we were sorting out the PACKAGE_RANK bits for OFI.

Reply:
Yes it has. However, if you want to support inter-job shmem operations, a better choice might be `PMIX_NODE_RANK` as it is unique across jobs. `PMIX_LOCAL_RANK` is only unique within a job as it is the local rank _within_ that app.

Note that `PMIX_NODE_RANK` can have "holes" in it over time as jobs come/go as we don't renumber the node ranks. In other words:

* app A comes first and has procs with node rank of 0,1,2 - these will have local rank of 0,1,2
* app B starts and its procs will get node rank 3,4 - these will have local rank of 0,1
* app A completes and goes away
* app C starts - its procs will get node rank 5,6,7 - these will have local rank of 0,1,2

So "node rank" monotonically increases until it hits `UINT16_MAX`, upon which it recycles back to take the lowest numbers again (won't go over the details here).

Reply:
So for now it sounds like PMIX_LOCAL_RANK is the right one given we still apparently base everything off the local job:

```
        /* check to see if this proc can be reached via shmem (i.e.,                                                                                                                                                                                                                 
           if they're on my local host and in my job) */
        if (procs[proc]->proc_name.jobid != my_proc->proc_name.jobid ||
            !OPAL_PROC_ON_LOCAL_NODE(procs[proc]->proc_flags)) {
            peers[proc] = NULL;
            continue;
        }
```

I can follow this up with a change to support multiple jobs or I can update this PR.

Reply:
The main thing I need to do is change how endpoints are allocated. Right now they are in one big block sized off the current job. I could change it to allocate an array of pointers instead. Adds some indirection but shouldn't hit latency too badly. I can test the change on the M1 where the latency is ~ 110 ns for the 1-byte ping pong.